magento/devdocs
# Replace glossary liquid tags

## Purpose of this pull request

This pull request (PR) replaces the old style of indicating glossary terms in the content with the standard markdown link.

Since this PR is a bulk change that affects a good amount of the source, it would be better to review the implementation and do smoke testing by randomly looking at some of the changed files.

## Implementation details

Tag replacement required VS code's bulk search and replace functionality and the following phases:

### Phase 1 - Find and replace liquid syntax

| Regex | Replacement |
| --- | --- |
| `\{%\s*glossarytooltip[\s\n]*[^%]+%}\n?([^{(%]+)\n?\{%[\s\n]*endglossarytooltip[\s\n]*%}` | `[$1](https://glossary.magento.com/$1)`|

Regex to find all `glossarytooltip` instances requires a complex regex pattern due to the many variants found in the source. E.g. `{%glossarytooltip%}`, `{% glossarytooltip%}`, `{% glossarytooltip %}`, etc.

### Phase 2 - Remove all spaces in the url location in the markdown link syntax

| Regex | Replacement|
| --- | --- |
|`glossary.magento.com/([^\s)]+) ([^\s)]+)\)` | `glossary.magento.com/$1-$2)`|
|`glossary.magento.com/([^\s)]+) ([^\s)]+) ([^\s)]+)\)`| `glossary.magento.com/$1-$2-$3)`|
|`glossary.magento.com/([^\s)]+) ([^\s)]+) ([^\s)]+) ([^\s)]+)\)` |`glossary.magento.com/$1-$2-$3-$4)`|

These expressions capture cases where there are 2, 3, and 4 space separated words in the url. There are no glossary terms that are more than 4 words.

### Phase 3 - Lowercase all uppercase characters

| Regex | Replacement |
| --- | --- |
| `glossary.magento.com/WYSIWYG` | `glossary.magento.com/wysiwyg`|
| `glossary.magento.com/XML` | `glossary.magento.com/xml`|
| `glossary.magento.com/URL` | `glossary.magento.com/url`|
| `glossary.magento.com/SKU` | `glossary.magento.com/sku`|
| `glossary.magento.com/PHTML` | `glossary.magento.com/phtml`|
| `glossary.magento.com/PHP` | `glossary.magento.com/php`|
| `glossary.magento.com/HTML` | `glossary.magento.com/html`|
| `glossary.magento.com/CSV` | `glossary.magento.com/csv`|
| `glossary.magento.com/CSS` | `glossary.magento.com/css`|
| `glossary.magento.com/CMS` | `glossary.magento.com/cms`|
| `glossary.magento.com/API` | `glossary.magento.com/api`|
| `glossary.magento.com/NGINX` | `glossary.magento.com/nginx`|
| `glossary.magento.com/CATALOG` | `glossary.magento.com/catalog`|
| `glossary.magento.com/A` | `glossary.magento.com/a`|
| `glossary.magento.com/B` | `glossary.magento.com/b`|
| `glossary.magento.com/C` | `glossary.magento.com/c`|
| `glossary.magento.com/D` | `glossary.magento.com/d`|
| `glossary.magento.com/E` | `glossary.magento.com/e`|
| `glossary.magento.com/F` | `glossary.magento.com/f`|
| `glossary.magento.com/G` | `glossary.magento.com/g`|
| `glossary.magento.com/JavaScript` | `glossary.magento.com/javascript`|
| `glossary.magento.com/JQuery` | `glossary.magento.com/jquery`|
| `glossary.magento.com/jQuery` | `glossary.magento.com/jquery`|
| `glossary.magento.com/J` | `glossary.magento.com/j`|
| `glossary.magento.com/L` | `glossary.magento.com/l`|
| `glossary.magento.com/M` | `glossary.magento.com/m`|
| `glossary.magento.com/N` | `glossary.magento.com/n`|
| `glossary.magento.com/O` | `glossary.magento.com/o`|
| `glossary.magento.com/P` | `glossary.magento.com/p`|
| `glossary.magento.com/S` | `glossary.magento.com/s`|
| `glossary.magento.com/T` | `glossary.magento.com/t`|
| `glossary.magento.com/UI` | `glossary.magento.com/ui`|
| `glossary.magento.com/W` | `glossary.magento.com/w`|
| `glossary.magento.com/([a-z-]+)API` | `glossary.magento.com/([a-z-]+)api`|
| `glossary.magento.com/([a-z-]+)A` | `glossary.magento.com/([a-z-]+)a`|
| `glossary.magento.com/([a-z-]+)B` | `glossary.magento.com/([a-z-]+)b`|
| `glossary.magento.com/([a-z-]+)C` | `glossary.magento.com/([a-z-]+)c`|
| `glossary.magento.com/([a-z-]+)Delivery-Network` | `glossary.magento.com/([a-z-]+)delivery-network`|
| `glossary.magento.com/([a-z-]+)Engine-Optimization` | `glossary.magento.com/([a-z-]+)engine-optimization`|
| `glossary.magento.com/([a-z-]+)F` | `glossary.magento.com/([a-z-]+)f`|
| `glossary.magento.com/([a-z-]+)I` | `glossary.magento.com/([a-z-]+)i`|
| `glossary.magento.com/([a-z-]+)M` | `glossary.magento.com/([a-z-]+)m`|
| `glossary.magento.com/([a-z-]+)P` | `glossary.magento.com/([a-z-]+)p`|
| `glossary.magento.com/([a-z-]+)R` | `glossary.magento.com/([a-z-]+)r`|
| `glossary.magento.com/([a-z-]+)S` | `glossary.magento.com/([a-z-]+)s`|
| `glossary.magento.com/([a-z-]+)V` | `glossary.magento.com/([a-z-]+)v`|

There was no quicker way(in VS Code) of bulk lowercasing capitalized letters in the URL besides specifying individual letters or terms.

## Validation steps

1. Search the entire project directory for `glossarytooltip`
2. Verify there is no instance of this tag in any **markdown** file.
3. Use the following regex to find ANY capitalized letters in a `glossary.magento.com URL` in the source: `glossary.magento.com/[a-z-]*[A-Z]+[a-z-]*`
4. Verify there are no results found.
5. Build the site with`rake`.
6. Verify the site builds
7. Navigate to a page with glossary terms.
8. Verify that the term looks similar to the previous tooltip and not a regular link.
9. Click on a glossary term
10. Verify you land on a valid glossary page 

## Known issues

* Since glossary terms are now in markdown, instances where glossary terms were found inside HTML tags will not render the markdown correctly (unless the tag contains the following property `markdown="1"`). Fixing these is probably out of the scope of this PR and will require follow up PR's by us or community members that run into these instances.
* The glossary term is rendered as blue(since it is a link) instead of black. Changing this requires a change to the base doc theme and is outside the scope of this PR and repo.
